File: 2015/day_10/part1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// solution time: 03:24.00
char* look_and_say(const char* input) {
    static char result[500000];
    char current_char = input[0];
    int count = 1;
    int j = 0;

    for (int i = 1; i < strlen(input); i++) {
        if (input[i] == current_char) {
            count++;
        } else {
            j += sprintf(&result[j], "%d%c", count, current_char);
            current_char = input[i];
            count = 1;
        }
    }
    sprintf(&result[j], "%d%c", count, current_char);

    return result;
}

int main() {
    // read input.in
    FILE *fp = fopen("input.in", "r");

    if (fp == NULL) {
        printf("Could not open file input.in\n");
        return 1;
    }

    char line[500000];

    if (fgets(line, sizeof(line), fp) == NULL) {
        printf("Could not read line from file input.in\n");
        return 1;
    }

    fclose(fp);

    // process input
    for (int i = 0; i < 40; i++) {
        strcpy(line, look_and_say(line));
    }

    printf("%ld", strlen(line));

    return 0;
}

File: 2016/day_02/part1.py
# open and read the file to a string
input = open("input.in", "r")
input = input.read()
input = input.split("\n")

# Solution Time: 1:05
keypad = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
x, y = 1, 1
directions = {'U': (-1, 0), 'D': (1, 0), 'L': (0, -1), 'R': (0, 1)}
bathroom_code = ""

for line in input:
    for instruction in line:
        dx, dy = directions[instruction]
        if 0 <= x + dx < 3 and 0 <= y + dy < 3:
            x += dx
            y += dy
    bathroom_code += str(keypad[x][y])

print(bathroom_code)
